### PR TITLE
Ag 5559 drive docs with ast

### PR DESCRIPTION
**Pull request by StephenCooper**
The aim here is to generate interfaces and type data from the code via AST parsing to provide accurate information to the documentation pages.

### CODE REVIEW

File: grid-packages/ag-grid-docs/documentation/doc-pages/grid-api/grid-options.json [hunk 1]
```diff
@@ -0,0 +1,1533 @@
+{
+  "suppressBrowserResizeObserver": {
+    "description": "*****************************************************************************************************\nIf you change the properties on this interface, you must also update PropertyKeys to be consistent. *\n*****************************************************************************************************",
```
Comment (StephenCooper) [inline grid-packages/ag-grid-docs/documentation/doc-pages/grid-api/grid-options.json:3]: I will need to correct this description parsing in the next stage. It is not used yet in the ApiDocumentation component.

File: grid-packages/ag-grid-docs/package.json [hunk 2]
```diff
@@ -13,7 +13,8 @@
     "validate-es5": "npx eslint documentation/doc-pages",
     "test": "npx jest && cd documentation && npm run test",
     "hash": "sh ../../scripts/hashDirectory.sh > .hash",
-    "testtt": "node dev-server.js"
+    "testtt": "node dev-server.js",
+    "update-properties": "node -e \"require('./updateGridAndColumnProperties.js').updatePropertiesBuilt()\""
```
Comment (StephenCooper) [inline grid-packages/ag-grid-docs/package.json:17]: Need to update the name of this file and task to be more accurate.

File: grid-packages/ag-grid-docs/documentation/src/components/ApiDocumentation.jsx [hunk 3]
```diff
@@ -387,40 +434,106 @@ const FunctionCodeSample = ({ framework, name, type }) => {
...
+    })
 
-    return lines;
+    return allLines;
 };
 
 const getLinkedType = (type, framework) => {
     if (!Array.isArray(type)) {
-        type = [type];
+        type = type.split('|').map(t => t ? t.trim() : t);
```
Comment (StephenCooper) [inline grid-packages/ag-grid-docs/documentation/src/components/ApiDocumentation.jsx:531]: Plan to use a regex here to improve this and make it cleaner.

File: grid-packages/ag-grid-docs/documentation/src/components/ApiDocumentation.jsx [hunk 4]
```diff
@@ -387,40 +434,106 @@ const FunctionCodeSample = ({ framework, name, type }) => {
...
+
+        } else {
+
+            // We have an interface
+            const lines = [`interface ${name} {`];
+
+            const properties = Object.entries(interfaceType.type);
+            properties.sort(([p1,], [p2,]) => {
+                // Sort alphabetically but with $scope at the end
+                if (p1 === '$scope') return 1;
```
Comment (StephenCooper) [inline grid-packages/ag-grid-docs/documentation/src/components/ApiDocumentation.jsx:512]: Any thoughts on hiding the $scope property from the Angular, React, Vue frameworks?
Or not worth removing now seeing as will soon be removed for real.

**Review by rmc-software — APPROVED** [review]

File: grid-packages/ag-grid-docs/documentation/src/components/ApiDocumentation.jsx (hunk 4 above)
Comment (seanlandsman) [reply grid-packages/ag-grid-docs/documentation/src/components/ApiDocumentation.jsx:511]: I don't think it'll be soon - it'll still be there in 26.x and as that'll be our LTS version it's unlikely to be removed entirely (from docs) soon

Comment (seanlandsman) [reply grid-packages/ag-grid-docs/documentation/src/components/ApiDocumentation.jsx:511]: but that said, I can't see why we'd display this in any context other than Angular (even though angular isn't angularjs...)

**Review by seanlandsman — COMMENTED** [review]
Looks fine to me - I've not really reviewed the actual AST changes in depth but the docs & resulting angular outputs look good as far as I can tell

File: grid-packages/ag-grid-docs/documentation/src/components/ApiDocumentation.jsx (hunk 4 above)
Comment (StephenCooper) [reply grid-packages/ag-grid-docs/documentation/src/components/ApiDocumentation.jsx:511]: Yeah, wasn't sure if would need it in both Javascript and Angular to catch users of AngularJS. 

But sounds like hiding it from Vue and React would make sense.

**Pull request merged**